File: 141-linked-list-cycle/linked-list-cycle.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public boolean hasCycle(ListNode head) 
    {
        if(head == null || head.next == null)
        {
            return false;
        }

        ListNode slow = head;
        ListNode fast = head.next;

        while((fast != null) && (fast != slow))
        {
            fast = fast.next;

            if(fast == null || fast == slow)
            {
                break;
            }

            slow = slow.next;
            fast = fast.next;
        }

        if(fast == null)
        {
            return false;
        }
        else
        {
            return true;
        } 
    }
}

File: 141-linked-list-cycle/linked-list-cycle.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        if head is None:
            return False

        slow = head
        fast = head.next

        while fast is not None and slow != fast:
            fast = fast.next.next if fast.next is not None else fast.next
            slow = slow.next

        if fast is None:
            return False
        else:
            return True